libraries/chain/finality/finalizer.cpp

```diff
@@ -85,6 +85,19 @@ finalizer::vote_result finalizer::decide_vote(const block_state_ptr& bsp) {
    return res;
 }
 
+// ----------------------------------------------------------------------------------------
+// finalizer has voted strong on bsp, update finalizer safety info if more recent than the current lock
+bool finalizer::maybe_update_fsi(const block_state_ptr& bsp) {
+   auto& final_on_strong_qc_block_ref = bsp->core.get_block_reference(bsp->core.final_on_strong_qc_block_num);
+   if (final_on_strong_qc_block_ref.timestamp > fsi.lock.timestamp) {
+      fsi.lock = { final_on_strong_qc_block_ref.block_id, final_on_strong_qc_block_ref.timestamp };
+      fsi.last_vote             = { bsp->id(), bsp->timestamp() };
```

Comment:
I thgink maybe we should update `fsi.last_vote` only if `fsi.last_vote.timestamp` < `bsp->timestamp()` to make sure we don't break the monotony check.